File: docs/JAMIA/comparisons/RPython/TM/18401112_string_labels_roc_auc.py
# ID: SO#18401112
# Language: Python
# Bug class: Value type
# Runtime consequence: Crash
# In study: Yes
# Inclusion rationale: String labels ('0','1') instead of int labels for roc_auc_score. Wrong type at data load.
# Reproduction status: Fixed (modern sklearn)
# Type system catch: `Record<string, number>` rejects string column

import numpy as np
from sklearn.metrics import roc_auc_score

y_scores = np.array([0.63, 0.53, 0.36, 0.02, 0.70, 1, 0.48, 0.46, 0.57])
y_true = np.array(['0', '1', '0', '0', '1', '1', '1', '1', '1'])  # strings, not ints

try:
    result = roc_auc_score(y_true, y_scores)
    print(f"Result: {result}")
except ValueError as e:
    print(f"Crash: {e}")

# Fix: use integer labels
y_true_int = np.array([0, 1, 0, 0, 1, 1, 1, 1, 1])
print(f"Fixed: {roc_auc_score(y_true_int, y_scores)}")

File: docs/JAMIA/comparisons/RPython/TM/18401112_string_labels_roc_auc.ts
/**
 * RPython SO#18401112 — ValueError: Data is not binary and pos_label is not specified
 * Effect: Crash
 * Bug class: Type coercion
 *
 * In Python, roc_auc_score crashes when labels are strings ('0','1') instead of
 * integers (0, 1). The type mismatch is only discovered at runtime.
 *
 * In tidy-ts, a column of string labels is typed as string[]. Any function
 * expecting number[] would produce a compile-time error.
 */
import { createDataFrame, stats as s } from "@tidy-ts/dataframe";

const df = createDataFrame([
  { score: 0.63, label: "0" },
  { score: 0.53, label: "1" },
  { score: 0.36, label: "0" },
  { score: 0.70, label: "1" },
  { score: 1.0, label: "1" },
]);

// @ts-expect-error — string[] is not assignable to number[]
const wrong = s.mean(df.extract("label"));

// Fix: explicit conversion
const numeric = df.mutate({ label: (r) => parseInt(r.label, 10) });
const labels = numeric.extract("label"); // number[]
console.log(labels);
